flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/MongoSourceEnumStateSerializer.java

```diff
@@ -0,0 +1,114 @@
+@Internal
+public class MongoSourceEnumStateSerializer
+        implements SimpleVersionedSerializer<MongoSourceEnumState> {
+
+    public static final MongoSourceEnumStateSerializer INSTANCE =
+            new MongoSourceEnumStateSerializer();
+
+    private MongoSourceEnumStateSerializer() {
+        // Singleton instance.
+    }
+
+    @Override
+    public int getVersion() {
+        // We use MongoSourceSplitSerializer's version because we use reuse this class.
+        return MongoSourceSplitSerializer.CURRENT_VERSION;
+    }
+
+    @Override
+    public byte[] serialize(MongoSourceEnumState state) throws IOException {
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                DataOutputStream out = new DataOutputStream(baos)) {
+            serializeList(out, state.getRemainingCollections(), DataOutputStream::writeUTF);
+
+            serializeList(out, state.getAlreadyProcessedCollections(), DataOutputStream::writeUTF);
+
+            serializeList(
+                    out,
+                    state.getRemainingScanSplits(),
+                    MongoSourceSplitSerializer.INSTANCE::serializeMongoSplit);
+
+            serializeMap(
+                    out,
+                    state.getAssignedScanSplits(),
+                    DataOutputStream::writeUTF,
+                    MongoSourceSplitSerializer.INSTANCE::serializeMongoSplit);
+
+            out.writeBoolean(state.isInitialized());
+
+            out.flush();
+            return baos.toByteArray();
+        }
+    }
+
+    @Override
+    public MongoSourceEnumState deserialize(int version, byte[] serialized) throws IOException {
+        try (ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
+                DataInputStream in = new DataInputStream(bais)) {
+            List<String> remainingCollections = deserializeList(in, DataInput::readUTF);
+            List<String> alreadyProcessedCollections = deserializeList(in, DataInput::readUTF);
+            List<MongoScanSourceSplit> remainingScanSplits =
+                    deserializeList(in, i -> deserializeMongoScanSourceSplit(version, i));
+
+            Map<String, MongoScanSourceSplit> assignedScanSplits =
+                    deserializeMap(
+                            in,
+                            DataInput::readUTF,
+                            i -> deserializeMongoScanSourceSplit(version, i));
+
+            boolean initialized = in.readBoolean();
+
+            return new MongoSourceEnumState(
+                    remainingCollections,
+                    alreadyProcessedCollections,
+                    remainingScanSplits,
+                    assignedScanSplits,
+                    initialized);
+        }
+    }
+
+    private MongoScanSourceSplit deserializeMongoScanSourceSplit(int version, DataInputStream in)
+            throws IOException {
+        return (MongoScanSourceSplit)
```

Comment:
MongoDB can support not only batch reads, but also streaming reads.
We can use MongoDB [change streams](https://www.mongodb.com/docs/manual/changeStreams/#change-streams) feature to read operation changes, and then convert them into flink's upsert mode changelog.

Batch read is currently implemented in this PR, some abstractions are also left for streaming reading:
MongoSourceSplit -> MongoScanSourceSplit
MongoSourceSplit -> MongoStreamSourceSplit

Here I made some adjustments in recent commit, check the split type to avoid type casting.
Whether we need to left some abstractions for this now, I would like to hear your opinion.